storescu/src/main.rs

```diff
@@ -164,7 +164,27 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
 
         scu.send(&pdu)?;
 
-        scu.send_pdata(pc_selected.id).write_all(&object_data)?;
```

Comment:
Hi @Enet4 yes, this is a re-implementation but much more simpler and explicit to read and verify. If you look carefully my patch is made so that except for the iniitial PDUs send to the peer during association establishment, all subsequent PDUs will be send with `scu.send()` which is a good thing.

IMHO, should not bother fixing [pdata.rs](https://github.com/Enet4/dicom-rs/blob/master/ul/src/association/pdata.rs), you should simply discard it completly.

It would be better in the long term to extract my re-implementation as a function taking as input a `Reader` and a max PDU length and returning an iterator yielding PDUs to be sent with `scu.send(&pdu).

With this in place, we could achieve some nice things, like in  `scpproxy`, taking big PDUs as input and splitting them in smaller chunks to be sent to a memory constrained, embedded PACS, or taking small PDUs as input and reassembling them in large chunks in the other way.

Regards, and congrats for this amazing piece of software !

Reply:
Thank you for the feedback, @charbeljc. I kept looking for the possible issue, and it turns out that the association client isn't reading the maximum PDU length from the association acknowledgement message. In the event that a device is using anything lower than 16384 as the maximum allowed PDU size, subsequent interactions may fail unexpectedly. I will file an issue once possible, and with this fixed, chances are that the other abstractions will also start working as intended.


> IMHO, you should not bother fixing pdata.rs, you should simply discard it completly.
>
> It would be better in the long term to extract my re-implementation as a function taking as input a `Reader` and a max PDU length and returning an iterator yielding PDUs to be sent with `scu.send(&pdu)`.

This will need more brainstorming before any decisions are made to replace abstractions already in place. For what it's worth, the current one can already automatically transform an arbitrary source of bytes into PDU transmissions with the intended length. It is only not apparent here because this program is writing the data set into a vector first, but that would not even be necessary. And I can picture a different abstraction for an iterator of PDUs, as you suggest, but unless I overlooked a better way, this would seemingly introduce more challenges and not necessarily more benefits than the one available in the latest version.